Adding option --no-deploy

r-f-g opened the pull request:
This will allow you to skip all functions marked with the `abort_on_fail`
marker, which represents bundle/charm deployment.

r-f-g commented:
Example of use:
```python
import pytest


@pytest.mark.abort_on_fail
async def test_build_and_deploy(ops_test):
    await ops_test.model.deploy("ubuntu")
    await ops_test.model.wait_for_idle(wait_for_active=True, timeout=60 * 60)


async def test_unit_status(ops_test):
    unit = ops_test.model.applications["ubuntu"].units[0]
    assert unit.workload_status == "active"
```
and [output](https://pastebin.ubuntu.com/p/xjpkhxGBRr/).

addyess commented:
@r-f-g, i like the idea of what you're going for here -- but i'm not sure using `pytest.mark.abort_on_fail` is the right tag for this behavior.  

[from the docs](https://github.com/charmed-kubernetes/pytest-operator/blob/main/docs/reference.md#pytestmarkasyncio)
> If the test marked with this fails, it will cause any subsequent test in the module to automatically xfail. This should be used when a given test is a requirement for any following tests to have any possibility of succeeding, such as if the test does the initial build and deploy.

Have you considered using pytest's conditional skipif markers?

```python
@pytest.mark.skipif(pytest.config.getvalue("no-deploy"), reason="--no-deploy was specified")
async def test_build_and_deploy(ops_test)
    await ops_test.model.deploy("ubuntu")
    await ops_test.model.wait_for_idle(wait_for_active=True, timeout=60 * 60)
````

addyess commented:
also, if we're adding a new config method, please update the [reference doc](https://github.com/charmed-kubernetes/pytest-operator/blob/main/docs/reference.md)

r-f-g commented:
> @r-f-g, i like the idea of what you're going for here -- but i'm not sure using pytest.mark.abort_on_fail is the right tag for this behavior.

In most of the cases I've seen, it was used for testing to deploy the model, so I assumed it would be quite appropriate.

> Have you considered using pytest's conditional skipif markers?

It could work like this:
```python
@pytest.mark.skipif(pytest.config.getvalue("no-deploy") and pytest.config.getvalue("model") is not None, reason="--no-deploy was specified")
```
However, its use will not be intuitive and could cause inconsistencies in different projects. What do you think about following approaches:
- add another custom `deployment` marker (it could also behave the same as `abort_on_fail`), e.g. `@pytest.mark.deployment`
- used `abort_on_fail` with arguments, e.g. `@pytest.mark.abort_on_fail(deploy=True)`

> also, if we're adding a new config method, please update the [reference doc](https://github.com/charmed-kubernetes/pytest-operator/blob/main/docs/reference.md)

Thanks, I overlooked the docs and also I should add there `--crash-dump` and `--crash-dump-output`.

addyess commented:
i'd be in favor of a new marker then that makes it easier to read.  If you write "skip_if_deployed` to check the "--no-deploy" and "--model" flags

```python
@pytest.mark.skip_if_deployed
async def test_build_and_deploy(ops_test):
   charm = await ops_test.build(".")
   await ops_test.deploy(charm)
```

r-f-g commented:
@addyess I added the `skip_if_deployed` marker and updated the docs. However, I stayed with the `pytest_runtest_setup` function, as it seems better to me than using fixtures here. The idea came from pytest [docs](https://github.com/pytest-dev/pytest/blob/0ad452bcbadffa2bb890c543c8b409d2df61c67d/doc/en/example/markers.rst#custom-marker-and-command-line-option-to-control-test-runs).

addyess reviewed (approved):
I like it -- really clean.  Take care of the lint issues and i'll merge and push to pypi. 
LGTM

ca-scribner commented on tests/unit/test_pytest_operator.py:
```diff
@@ -291,3 +291,34 @@ async def mock_run(*cmd):
+    result.assert_outcomes(passed=3)
+
+    # test with --no-deploy and --model
+    result = testdir.runpytest("--no-deploy", "--model", "test-model")
```
related to comment above, can add a test here checking behaviour when we do `--no-deploy` without `--model`

ca-scribner commented on pytest_operator/plugin.py:
```diff
@@ -98,6 +105,15 @@ def pytest_configure(config):
     )
 
 
+def pytest_runtest_setup(item):
```
I think the name `pytest_runtest_setup` inherited from [here](https://github.com/pytest-dev/pytest/blob/0ad452bcbadffa2bb890c543c8b409d2df61c67d/doc/en/example/markers.rst#custom-marker-and-command-line-option-to-control-test-runs) can be tailored to our use.  I believe it is called `runtest_setup` because it is a setting to decide if it runs tests (see the helps/comments in the example).  Ours would be slightly different (something to do with skipping if deployed?).

If we had two different conditions (say `skip_if_deployed` and `run_if_XXX`), would we implement both in this function or two separate functions?  I don't understand exactly how this works so am not sure, but that would influence how it should be named (and maybe a docstring explaining what it does would help too)

r-f-g replied:
> If we had two different conditions (say skip_if_deployed and run_if_XXX), would we implement both in this function or two separate functions? I don't understand exactly how this works so am not sure, but that would influence how it should be named (and maybe a docstring explaining what it does would help too)


Both should be in this function. There is this [runtest protocol](https://docs.pytest.org/en/7.0.x/reference/reference.html?highlight=pytest_runtest_setup#pytest.hookspec.pytest_runtest_protocol), where the `pytest_runtest_setup` is in from setup phase. I understand that it is similar to `setUp` function from `unittest.TestCase`.

I will add docstring to this function with explanation and reference to docs + examples.

ca-scribner commented:
Yeah the more I think about this, the more I like that it allows the author of tests to convey intent around what is deploy vs what is test.  At first I wondered if the `deploy` vs `not deploy` concepts were fundamental enough to enshrine them with marks and all that, but the more I think about it I think they are.  And there's nothing stopping someone from still using `-k` or other things - this just adds a convenience

r-f-g commented on pytest_operator/plugin.py:
```diff
@@ -98,6 +105,15 @@ def pytest_configure(config):
 
+def pytest_runtest_setup(item):
+    if (
+        "skip_if_deployed" in item.keywords
+        and item.config.getoption("--no-deploy")
+        and item.config.getoption("--model") is not None
+    ):
```
This is really good idea. But I only found this "ugly" solution.
```python
def pytest_addoption(parser):
    ...
    arg_parser = parser._getparser()
    args = arg_parser.parse_args()
    if args.no_deploy and args.model is None:
        arg_parser.error("model is required")
```

addyess replied:
doesn't this work:


```suggestion
        item.config.getoption("--no-deploy")
        and item.config.getoption("--model") is None
    ):
        pytest.exit("--model `` is required when --no-deploy is specified")

    if (
        "skip_if_deployed" in item.keywords
        and item.config.getoption("--no-deploy")
        and item.config.getoption("--model") is not None
    ):
```

r-f-g replied:
That works, but the output is not really nice.
```bash
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! _pytest.outcomes.Exit: --model is required when --no-deploy is specified !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================================================================================================= 2 warnings in 0.19s ==================================================================================================================================================
ERROR: InvocationError for command /home/rg/code/canonical/test-pytest-operator/.tox/integration/bin/pytest --tb native --show-capture=no --log-cli-level=INFO -s --no-deploy tests/integration (exited with code 2)
_______________________________________________________________________________________________________________________________________________________ summary ________________________________________________________________________________________________________________________________________________________
ERROR:   integration: commands failed
```

addyess replied:
¯\_(ツ)_/¯

i see whatcha mean

r-f-g replied:
@addyess @ca-scribner Do you think it's okay to go with that "ugly" solution?

ca-scribner replied:
Sorry for being so slow in response.  The ugly solution was the best I had in my head, too.  I don't think it is too bad, and it can at least be very clear in the message to the user.  Only nit I'd add is to add to the error message something about --no-deploy, like "must specify --model when using --no-deploy" or similar

r-f-g replied:
Ok, I implemented it and here is an [example](https://pastebin.ubuntu.com/p/pKmX86XPJV/) of how it works.
I struggled with unittests a bit, but in the end I found (in the pytest source code) `runpytest_subprocess` as alternative to `runpytest`.

r-f-g replied:
Your suggestion is about the error message `must specify --model when using --no-deploy` or do you suggesting to use `pytest.exit` in `pytest_runtest_setup`?

addyess replied:
oh @r-f-g I can get behind your current solution.  Okay to mark this thread as "resolved"?

r-f-g replied:
For me, yes. I'm not sure if @ca-scribner has other suggestions.

ca-scribner replied:
lgtm!

ca-scribner commented:
I think this is now good to go?

r-f-g commented:
@addyess could we re-run tests?

addyess commented:
@r-f-g let's rebase this too
